File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.java
class Solution {
    public int sumSubarrayMins(int[] arr) {
        
        int n = arr.length;
        int[] nsl = new int[n];
        
        Stack<Integer> st = new Stack<>();
        for(int i=0; i<n; i++){
            while(st.size()>0 && arr[st.peek()] >= arr[i]) st.pop();
            
            if(st.size()==0) nsl[i] = -1;
            else nsl[i] = st.peek();
            
            st.push(i);
        }
        
        st.clear();
        
        int[] nsr = new int[n];
        for(int i=n-1; i>=0; i--){
            while(st.size()>0 && arr[st.peek()] > arr[i]) st.pop();
            
            if(st.size()==0) nsr[i] = n;
            else nsr[i] = st.peek();
            
            st.push(i);
        }
        
        long sum=0;
        int MOD = 1000000007;
        
        for(int i=0; i<n; i++){
            long product = (long)arr[i] * ((long)(i-nsl[i])*(nsr[i]-i));
            sum = (sum % MOD + product % MOD) % MOD;
        }
        
        return (int)sum;
    }
}

File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.py
class Solution:
    def sumSubarrayMins(self, arr: List[int]) -> int:
        
        LENGTH = len(arr)
        stack = []
        nsl = [ -1 for x in range(LENGTH)]
        
        for i in range(LENGTH):
            while len(stack)>0 and arr[stack[-1]] >= arr[i]:
                stack.pop()
            
            if len(stack)!=0:
                nsl[i] = stack[-1]
            
            stack.append(i)
        
        stack.clear()
        nsr=[LENGTH for x in range(LENGTH)]
        
        for i in range(LENGTH-1, -1, -1):
            while len(stack)>0 and arr[stack[-1]] > arr[i]:
                stack.pop()
            
            if len(stack)!=0:
                nsr[i] = stack[-1]
            
            stack.append(i)
        
        print(nsl, nsr)
        total=0
        MOD = 10**9+7
        for i in range(len(arr)):
            product = (arr[i])*((i-nsl[i])*(nsr[i]-i))
            total = ((total % MOD) + (product % MOD))  % MOD
        
        return total